src/imex/transfer.rs

```diff
@@ -94,267 +90,181 @@ impl BackupProvider {
-                    break Err(anyhow!("BackupProvider cancelled"));
-                },
+                    context.emit_event(EventType::ImexProgress(0));
+                    break;
+                }
                 _ = drop_token.cancelled() => {
-                    provider.shutdown();
-                    break Err(anyhow!("BackupProvider dropped"));
+                    context.emit_event(EventType::ImexProgress(0));
+                    break;
                 }
             }
-        };
-        match &res {
-            Ok(_) => {
-                context.emit_event(SendProgress::Completed.into());
-                let mut msg = Message::new(Viewtype::Text);
-                msg.text = backup_transfer_msg_body(context).await;
-                add_device_msg(context, None, Some(&mut msg)).await?;
-            }
-            Err(err) => {
-                error!(context, "Backup transfer failure: {err:#}");
-                context.emit_event(SendProgress::Failed.into())
-            }
         }
-        res
     }
 
     /// Returns a QR code that allows fetching this backup.
     ///
     /// This QR code can be passed to [`get_backup`] on a (different) device.
     pub fn qr(&self) -> Qr {
-        Qr::Backup {
-            ticket: self.ticket.clone(),
-        }
-    }
-}
+        Qr::Backup2 {
+            node_addr: self.node_addr.clone(),
 
-impl Future for BackupProvider {
-    type Output = Result<()>;
-
-    fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
-        Pin::new(&mut self.handle).poll(cx)?
+            auth_token: self.auth_token.clone(),
+        }
     }
-}
-
-/// Create [`EventType::ImexProgress`] events using readable names.
-///
-/// Plus you get warnings if you don't use all variants.
-#[derive(Debug)]
-enum SendProgress {
-    Failed,
-    Started,
-    DatabaseExported,
-    CollectionCreated,
-    ProviderListening,
-    ClientConnected,
-    TransferInProgress { current_size: u64, total_size: u64 },
-    Completed,
-}
 
-impl From<SendProgress> for EventType {
-    fn from(source: SendProgress) -> Self {
-        use SendProgress::*;
-        let num: u16 = match source {
-            Failed => 0,
-            Started => 100,
-            DatabaseExported => 300,
-            CollectionCreated => 350,
-            ProviderListening => 400,
-            ClientConnected => 450,
-            TransferInProgress {
-                current_size,
-                total_size,
-            } => {
-                // the range is 450..=950
-                450 + ((current_size as f64 / total_size as f64) * 500.).floor() as u16
-            }
-            Completed => 1000,
-        };
-        Self::ImexProgress(num.into())
+    /// Waits for the backup transfer to complete.
+    pub async fn wait(&mut self) -> Result<()> {
+        let handle = tokio::spawn(async { Ok(()) });
+        let old_handle = std::mem::replace(&mut self.handle, handle);
```

Comment:
This is not ideal state-wise.  I think this is why it was implementing future directly before.